### PR TITLE
Add 'Tracker' track-generation tool

### PR DESCRIPTION
**Pull request by drawlerr**
Tracker is a small Python tool for automatically generating a new Rally track from an existing repository

Relates #444

### CODE REVIEW

File: tracker/corpus.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+                comp_outfile.write(compressor.compress(data))
+
+                render_progress(n+1, total_docs, freq)
+
+            print()  # progress prints didn't have a newline
+            comp_outfile.write(compressor.flush())
+    return template_vars(index, outpath, comp_outpath, total_docs)
+
+
+def render_progress(cur, total, freq):
```
Comment (danielmitterdorfer) [inline tracker/corpus.py]: Can we use Rally's progress bar here instead? (In general, output to the terminal should be done via `console`).

File: tracker/templates/challenges.json.j2 [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+          "operation": {
+            "operation-type": "create-index",
+            "settings": {% raw %}{{index_settings | default({}) | tojson}}{% endraw %}
+          }
+        },
+        {
+          "operation": "index-append",
+          "warmup-time-period": 120,
+          "clients": {% raw %}{{bulk_indexing_clients | default(8)}}{% endraw %}
+        },
```
Comment (danielmitterdorfer) [inline tracker/templates/challenges.json.j2]: IMHO it's ok to drop every task below this line for the initial track template.

File: tracker/templates/challenges.json.j2 [hunk 3]
```diff
@@ -0,0 +1,49 @@
+    {
+      "name": "append-no-conflicts",
+      "description": "Indexes the whole document corpus using Elasticsearch default settings. We only adjust the number of replicas as we benchmark a single node cluster and Rally will only start the benchmark if the cluster turns green. Document ids are unique so all index operations are append only. After that a couple of queries are run.",
```
Comment (danielmitterdorfer) [inline tracker/templates/challenges.json.j2]: The description does not match what's going on.

File: tracker/templates/operations.json.j2 [hunk 4]
```diff
@@ -0,0 +1,16 @@
+{
+    "name": "index-append",
+    "operation-type": "bulk",
+    "bulk-size": {{bulk_size | default(5000)}},
+    "ingest-percentage": {{ingest_percentage | default(100)}}
+},
+{
+    "name": "index-update",
```
Comment (danielmitterdorfer) [inline tracker/templates/operations.json.j2]: Unused?

File: tracker/templates/track.json.j2 [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+          "target-index": "{{corpus.index_name}}",
+          "source-file": "{{corpus.filename}}",
+          "document-count": {{corpus.doc_count}},
+          "compressed-bytes": {{corpus.compressed_bytes}},
+          "uncompressed-bytes": {{corpus.uncompressed_bytes}}
+        }{% if not loop.last %},{% endif %}{% endfor %}
+      ]
+    }
+  ],{% raw %}
+  "operations": [
```
Comment (danielmitterdorfer) [inline tracker/templates/track.json.j2]: IMHO we should use the simplest possible configuration and should define operations inline and only allow one challenge (see also our first example in the [Rally docs](https://esrally.readthedocs.io/en/stable/adding_tracks.html)). This is not as maintainable but also much simpler to get started.

File: tracker/tracker.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    log_conf = load_json(log_config_path)
+    logging.config.dictConfig(log_conf)
+    logging.captureWarnings(True)
+
+
+def main():
+    configure_logging()
+    args = get_args()
+
+    client = Elasticsearch(hosts=args.hosts)
```
Comment (danielmitterdorfer) [inline tracker/tracker.py]: How about we use Rally's Elasticsearch client? We could then support all the options that Rally allows as well  (authentication etc.)?

File: tracker/corpus.py (hunk 1 above)
Comment (drawlerr) [reply tracker/corpus.py]: Sure, leftover from when Tracker was in its own repo.  Now that it's back in the Rally repo, this makes perfect sense 😃

File: tracker/tracker.py (hunk 6 above)
Comment (drawlerr) [reply tracker/tracker.py]: Addressed in https://github.com/elastic/rally/pull/875/commits/0dfdbfd3b501aabf3d09747745cd12a6acf5ddc4

File: tracker/templates/track.json.j2 (hunk 5 above)
Comment (drawlerr) [reply tracker/templates/track.json.j2]: Addressed in https://github.com/elastic/rally/pull/875/commits/ff4283398bdce2db2d7eb81158d6f39942347d2f

File: tracker/templates/operations.json.j2 (hunk 4 above)
Comment (drawlerr) [reply tracker/templates/operations.json.j2]: I believe so, removed in https://github.com/elastic/rally/pull/875/commits/ff4283398bdce2db2d7eb81158d6f39942347d2f

File: tracker/corpus.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    total_docs = client.count(index=index)["count"]
+    logging.info("%d total docs in index %s", total_docs, index)
+    freq = total_docs // 1000
+
+    compressor = bz2.BZ2Compressor()
+    comp_outpath = outpath + ".bz2"
+
+    with open(outpath, "wb") as outfile:
+        with open(comp_outpath, "wb") as comp_outfile:
+            logging.info("Now dumping corpus to %s...", outpath)
```
[the remark this answers is not in the document]
Comment (drawlerr) [reply tracker/corpus.py]: Addressed in: https://github.com/elastic/rally/pull/875/commits/3d32262c4262e80321555af970bba2c65b794fad

File: tracker/corpus.py (hunk 1 above)
Comment (drawlerr) [reply tracker/corpus.py]: Addressed in: https://github.com/elastic/rally/pull/875/commits/3d32262c4262e80321555af970bba2c65b794fad

File: tracker/corpus.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+# specific language governing permissions and limitations
+# under the License.
+
+import bz2
+import json
+import logging
+import os
+import pathlib
+
+from elasticsearch import helpers
```
[the remark this answers is not in the document]
Comment (drawlerr) [reply tracker/corpus.py]: Addressed in https://github.com/elastic/rally/pull/875/commits/3d32262c4262e80321555af970bba2c65b794fad

File: setup.py [hunk 9]
```diff
@@ -118,7 +118,8 @@ def str_from_file(name):
       entry_points={
           "console_scripts": [
               "esrally=esrally.rally:main",
-              "esrallyd=esrally.rallyd:main"
+              "esrallyd=esrally.rallyd:main",
+              "tracker=tracker.tracker:main"
```
[the remark this answers is not in the document]
Comment (drawlerr) [reply setup.py]: Addressed in: https://github.com/elastic/rally/pull/875/commits/ef36af2d13a7639976f5a9f820e2b8253b023b62

File: tracker/templates/challenges.json.j2 (hunk 3 above)
Comment (drawlerr) [reply tracker/templates/challenges.json.j2]: https://github.com/elastic/rally/pull/875/commits/7d47ae90c56a03726cb7fccdeec21c2b7b36f171

File: tracker/templates/challenges.json.j2 (hunk 2 above)
Comment (drawlerr) [reply tracker/templates/challenges.json.j2]: https://github.com/elastic/rally/commit/1ef7f2addd8fae5833c82cdf502b317594aa6d80

**Comment by dliappis** [comment]
@elasticmachine test this

**Comment by dliappis** [comment]
@elasticmachine run tests

File: tracker/corpus.py [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+def template_vars(index_name, out_path, comp_outpath, doc_count):
+    corpus_path = pathlib.Path(out_path)
+    compressed_corpus_path = pathlib.Path(comp_outpath)
+    return {
+        "index_name": index_name,
+        "base_url": corpus_path.parent.as_uri(),
+        "filename": corpus_path.name,
+        "path": corpus_path,
+        "doc_count": doc_count,
+        "uncompressed_bytes": os.stat(corpus_path.as_posix()).st_size,
```
Comment (danielmitterdorfer) [inline tracker/corpus.py]: Can we use `os.path.getsize` instead here?

**Comment by drawlerr** [comment]
That was quite a review!  

The changes (including integration testing) will take more than one shot so I'll come back and ping you when I'm ready for another review.

**Comment by danielmitterdorfer** [comment]
Sure! Splitting the follow-up work into smaller items totally makes sense. :)

File: setup.py [hunk 11]
```diff
@@ -141,7 +141,8 @@ def str_from_file(name):
       entry_points={
           "console_scripts": [
               "esrally=esrally.rally:main",
-              "esrallyd=esrally.rallyd:main"
+              "esrallyd=esrally.rallyd:main",
+              "estracker=tracker.tracker:main"
```
[the remark this answers is not in the document]
Comment (drawlerr) [reply setup.py]: Changed in https://github.com/elastic/rally/pull/875/commits/8f9e846d56a09f00ec12ffd0575f434d99b9c927

File: tracker/tracker.py [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+    parser = argparse.ArgumentParser(description="Dump mappings and document-sources from an Elasticsearch index to create a Rally track.")
+
+    parser.add_argument("--target-hosts", default="", required=True, help="Elasticsearch host(s) to connect to")
+    parser.add_argument("--client-options", default=opts.ClientOptions.DEFAULT_CLIENT_OPTIONS, help="Elasticsearch client options")
+    parser.add_argument("--indices", nargs='+', required=True, help="Indices to include in track")
+    parser.add_argument("--trackname", help="Name of track to use, if different from the name of index")
+    parser.add_argument("--outdir", help="Output directory for track (default: tracks/)")
+
+    args = parser.parse_args()
+    if not args.trackname:
```
[the remark this answers is not in the document]
Comment (drawlerr) [reply tracker/tracker.py]: Changed in https://github.com/elastic/rally/pull/875/commits/0848be0aef5a3232af0078d456b1cbd823585f27

File: tracker/tracker.py [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+        outfile.write(template.render(template_vars))
+
+
+def get_args():
+    parser = argparse.ArgumentParser(description="Dump mappings and document-sources from an Elasticsearch index to create a Rally track.")
+
+    parser.add_argument("--target-hosts", default="", required=True, help="Elasticsearch host(s) to connect to")
+    parser.add_argument("--client-options", default=opts.ClientOptions.DEFAULT_CLIENT_OPTIONS, help="Elasticsearch client options")
+    parser.add_argument("--indices", nargs='+', required=True, help="Indices to include in track")
+    parser.add_argument("--trackname", help="Name of track to use, if different from the name of index")
```
[the remark this answers is not in the document]
Comment (drawlerr) [reply tracker/tracker.py]: Changed in https://github.com/elastic/rally/pull/875/commits/0848be0aef5a3232af0078d456b1cbd823585f27

File: tracker/tracker.py [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+    with open(dest_path, "w") as outfile:
+        outfile.write(template.render(template_vars))
+
+
+def get_args():
+    parser = argparse.ArgumentParser(description="Dump mappings and document-sources from an Elasticsearch index to create a Rally track.")
+
+    parser.add_argument("--target-hosts", default="", required=True, help="Elasticsearch host(s) to connect to")
+    parser.add_argument("--client-options", default=opts.ClientOptions.DEFAULT_CLIENT_OPTIONS, help="Elasticsearch client options")
+    parser.add_argument("--indices", nargs='+', required=True, help="Indices to include in track")
```
[the remark this answers is not in the document]
Comment (drawlerr) [reply tracker/tracker.py]: Hopefully I got most of them in https://github.com/elastic/rally/pull/875/commits/0848be0aef5a3232af0078d456b1cbd823585f27

File: tracker/tracker.py [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+    configure_logging()
+    args = get_args()
+
+    target_hosts = opts.TargetHosts(args.target_hosts)
+    client_options = opts.ClientOptions(args.client_options, target_hosts=target_hosts)
+    client = EsClientFactory(hosts=target_hosts.all_hosts[opts.TargetHosts.DEFAULT],
+                             client_options=client_options.all_client_options[opts.TargetHosts.DEFAULT]).create()
+
+    info = client.info()
+    logging.info("Connected to Elasticsearch %s version %s", info['name'], info['version']['number'])
```
[the remark this answers is not in the document]
Comment (drawlerr) [reply tracker/tracker.py]: Done in https://github.com/elastic/rally/pull/875/commits/0848be0aef5a3232af0078d456b1cbd823585f27

File: tracker/tracker.py [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+
+    outpath = os.path.join(args.outdir, args.trackname)
+    if not os.path.exists(outpath):
```
[the remark this answers is not in the document]
Comment (drawlerr) [reply tracker/tracker.py]: Done in https://github.com/elastic/rally/pull/875/commits/0848be0aef5a3232af0078d456b1cbd823585f27

File: tracker/tracker.py [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+    "challenges.json.j2": "challenges/default.json",
+}
+
+
+def process_template(template_filename, template_vars, dest_path):
+    env = Environment(loader=PackageLoader("tracker", "templates"))
+    template = env.get_template(template_filename)
+
+    parent_dir = pathlib.Path(dest_path).parent
+    if not parent_dir.exists():
```
[the remark this answers is not in the document]
Comment (drawlerr) [reply tracker/tracker.py]: Done in https://github.com/elastic/rally/pull/875/commits/0848be0aef5a3232af0078d456b1cbd823585f27

File: tracker/templates/challenges.json.j2 [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+                "name": "index-append",
+                "operation-type": "bulk",
+                "bulk-size": {{bulk_size | default(5000)}},
+                "ingest-percentage": {{ingest_percentage | default(100)}}
+          },
+          "warmup-time-period": 120,
+          "clients": {% raw %}{{bulk_indexing_clients | default(8)}}{% endraw %}
+        }
+      ]
+    }
```
[the remark this answers is not in the document]
Comment (drawlerr) [reply tracker/templates/challenges.json.j2]: Done in https://github.com/elastic/rally/pull/875/commits/0848be0aef5a3232af0078d456b1cbd823585f27

File: tracker/templates/track.json.j2 [hunk 19]
```diff
@@ -0,0 +1,29 @@
+{% raw %}{% import "rally.helpers" as rally with context %}{% endraw %}
+{
+  "version": 2,
+  "description": "Tracker-generated track for {{track_name}}",
+  "indices": [{% for index in indices %}
+    {
+      "name": "{{index.name}}",
+      "body": "{{index.filename}}"
+    }{% if not loop.last %},{% endif %}{% endfor %}
```
[the remark this answers is not in the document]
Comment (drawlerr) [reply tracker/templates/track.json.j2]: Addressed in https://github.com/elastic/rally/pull/875/commits/0848be0aef5a3232af0078d456b1cbd823585f27

File: tracker/tracker.py [hunk 20]
```diff
@@ -0,0 +1,119 @@
...
+
+    template_vars.update({
+        "indices": indices,
+        "corpora": corpora
+    })
+    for template, dest_filename in TRACK_TEMPLATES.items():
+        dest_path = os.path.join(outpath, dest_filename)
+        process_template(template, template_vars, dest_path)
+
+    logging.info("%s has been tracked!", args.trackname)
```
[the remark this answers is not in the document]
Comment (drawlerr) [reply tracker/tracker.py]: Added in https://github.com/elastic/rally/pull/875/commits/0848be0aef5a3232af0078d456b1cbd823585f27

File: tracker/templates/challenges.json.j2 [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+          "operation": {
+            "operation-type": "create-index",
+            "settings": {% raw %}{{index_settings | default({}) | tojson}}{% endraw %}
+          }
+        },
+        {
+          "operation": {
+                "name": "index-append",
+                "operation-type": "bulk",
+                "bulk-size": {{bulk_size | default(5000)}},
```
[the remark this answers is not in the document]
Comment (drawlerr) [reply tracker/templates/challenges.json.j2]: https://github.com/elastic/rally/pull/875/commits/0848be0aef5a3232af0078d456b1cbd823585f27

File: tracker/templates/challenges.json.j2 [hunk 22]
```diff
@@ -0,0 +1,26 @@
...
+                "ingest-percentage": {{ingest_percentage | default(100)}}
```
[the remark this answers is not in the document]
Comment (drawlerr) [reply tracker/templates/challenges.json.j2]: https://github.com/elastic/rally/pull/875/commits/0848be0aef5a3232af0078d456b1cbd823585f27

File: tracker/templates/track.json.j2 [hunk 23]
```diff
@@ -0,0 +1,29 @@
...
+      "name": "{{track_name}}",
+      "base-url": "{{corpora[0].base_url}}",
+      "documents": [{% for corpus in corpora %}
+        {
+          "target-index": "{{corpus.index_name}}",
+          "source-file": "{{corpus.filename}}",
+          "document-count": {{corpus.doc_count}},
+          "compressed-bytes": {{corpus.compressed_bytes}},
+          "uncompressed-bytes": {{corpus.uncompressed_bytes}}
+        }{% if not loop.last %},{% endif %}{% endfor %}
```
[the remark this answers is not in the document]
Comment (drawlerr) [reply tracker/templates/track.json.j2]: https://github.com/elastic/rally/pull/875/commits/0848be0aef5a3232af0078d456b1cbd823585f27

File: tracker/corpus.py [hunk 24]
```diff
@@ -0,0 +1,84 @@
...
+    :param client: Elasitcsearch client to scroll
+    :param outdir: Destination directory for corpus dump
+    :param index: Name of index to dump
+    :return: dict of properties describing the corpus for templates
+    """
+    from elasticsearch import helpers
+    outpath = os.path.join(outdir, "{}-documents.json".format(index))
+
+    total_docs = client.count(index=index)["count"]
+    logging.info("%d total docs in index %s", total_docs, index)
```
[the remark this answers is not in the document]
Comment (drawlerr) [reply tracker/corpus.py]: https://github.com/elastic/rally/pull/875/commits/0848be0aef5a3232af0078d456b1cbd823585f27

File: tracker/corpus.py (hunk 10 above)
Comment (drawlerr) [reply tracker/corpus.py]: https://github.com/elastic/rally/pull/875/commits/0848be0aef5a3232af0078d456b1cbd823585f27

File: docs/tracker.rst [hunk 25]
```diff
@@ -0,0 +1,36 @@
...
+Intro
+-----
+Tracker is a Rally utility for generating a Rally track template + corpus from one or more existing Elasticsearch
+indices.
+
+Usage
+-----
+To generate a new track, invoke Tracker as follows (replacing the example parameters below)::
+
+    estracker --target-hosts=host:9200,host2:9200 --indices=index_1 index_2 --trackname=xtrack --outdir=mytracks
```
Comment (dliappis) [inline docs/tracker.rst]: We should also document the `--client-options` e.g. with an example; scrapping real Cloud clusters will be THE most common use case.

Comment (dliappis) [reply docs/tracker.rst]: Also there's a typo here it's `--track-name` not `--trackname`

Comment (dliappis) [reply docs/tracker.rst]: Additionally the values for `--indices` should be enclosed in double quotes when dealing with multiple indices.

File: tracker/tracker.py [hunk 26]
```diff
@@ -0,0 +1,119 @@
...
+    indices = []
+    corpora = []
+    for index_name in args.indices:
+        index_vars = index.extract(client, outpath, index_name)
+        indices.append(index_vars)
+
+        corpus_vars = corpus.extract(client, outpath, index_name)
+        corpora.append(corpus_vars)
+
+    template_vars.update({
```
[the remark this answers is not in the document]
Comment (drawlerr) [reply tracker/tracker.py]: Addressed in https://github.com/elastic/rally/pull/875/commits/4be9f7246d103fb2a7e76e8abdf423ea25cfef3d

**Comment by drawlerr** [comment]
Looks like CI was aborted for Jenkins upgrade.

@elasticmachine run tests

File: tracker/templates/track.json.j2 [hunk 27]
```diff
@@ -0,0 +1,29 @@
...
+  "indices": [{% for index in indices %}
+    {
+      "name": "{{index.name}}",
+      "body": "{{index.filename}}"
+    }{% if not loop.last %},{% endif %}{% endfor %}
+  ],
+  "corpora": [
+    {
+      "name": "{{track_name}}",
+      "base-url": "{{corpora[0].base_url}}",
```
[the remark this answers is not in the document]
Comment (drawlerr) [reply tracker/templates/track.json.j2]: Addressed in: https://github.com/elastic/rally/pull/875/commits/46dc2c5c8fd4ae7634cf96097b1e649bcb6467a4

File: tracker/index.py [hunk 28]
```diff
@@ -0,0 +1,72 @@
...
+
+
+def filter_ephemeral_index_settings(settings):
+    """
+    Some of the 'settings' reported by Elasticsearch for an index are
+    ephemeral values, not useful for re-creating the index.
+    :param settings: Index settings reported by index.get()
+    :return: settings with ephemeral keys removed
+    """
+    return {k: v for k, v in settings.items() if k not in INDEX_SETTINGS_EPHEMERAL_KEYS}
```
[the remark this answers is not in the document]
Comment (drawlerr) [reply tracker/index.py]: Addressed in: https://github.com/elastic/rally/pull/875/commits/796c68f7db644bedb85b30eb9870fff948f36568

File: tracker/index.py [hunk 29]
```diff
@@ -0,0 +1,72 @@
...
+
+def extract(client, outdir, index):
+    """
+    Request index information to format in "index.json" for Rally
+    :param client: Elasticsearch client
+    :param outdir: destination directory
+    :param index: name of index
+    :return: None
+    """
+    filename = index + ".json"
```
[the remark this answers is not in the document]
Comment (drawlerr) [reply tracker/index.py]: Addressed in: https://github.com/elastic/rally/commit/796c68f7db644bedb85b30eb9870fff948f36568

File: tracker/index.py [hunk 30]
```diff
@@ -0,0 +1,72 @@
...
+    return {"mappings": mappings, "settings": {"index": index_settings}}
+
+
+def extract(client, outdir, index):
+    """
+    Request index information to format in "index.json" for Rally
+    :param client: Elasticsearch client
+    :param outdir: destination directory
+    :param index: name of index
+    :return: None
```
[the remark this answers is not in the document]
Comment (drawlerr) [reply tracker/index.py]: Addressed in: https://github.com/elastic/rally/commit/796c68f7db644bedb85b30eb9870fff948f36568

File: tracker/resources/logging.json [hunk 31]
```diff
@@ -0,0 +1,38 @@
...
+  "version": 1,
+  "formatters": {
+    "normal": {
+      "format": "%(asctime)s,%(msecs)d PID:%(process)d %(name)s %(levelname)s %(message)s",
+      "datefmt": "%Y-%m-%d %H:%M:%S",
+      "()": "esrally.log.configure_utc_formatter"
+    }
+  },
+  "handlers": {
+    "default_log_handler": {
```
[the remark this answers is not in the document]
Comment (drawlerr) [reply tracker/resources/logging.json]: Addressed in: https://github.com/elastic/rally/commit/796c68f7db644bedb85b30eb9870fff948f36568

File: docs/tracker.rst (hunk 25 above)
Comment (drawlerr) [reply docs/tracker.rst]: track-name and client-options: https://github.com/elastic/rally/pull/875/commits/59e9bb311c4a6b60a351cc699a156f585e46d49c

csv_to_list indices handling: https://github.com/elastic/rally/pull/875/commits/f962abf5f4b7fc84a449333e0617caa322fd7aaf

**Comment by danielmitterdorfer** [comment]
Closing in favor of https://github.com/elastic/rally/pull/970 (which is based on this PR)

**Pull request closed without merging**